Initial server

Walker60 opened the pull request.

Walker60 committed:
Add files via upload

Walker60 committed:
Add files via upload

Walker60 committed:
added database

ianfisk commented on server.py:
```diff
@@ -0,0 +1,23 @@
+def download(charts):
+
+    try:
+        return send_from_directory(directory=charts, filename=charts+'.jpeg')
```
@Walker60 you should probably also commit this `charts` directory if you have it so this script would not crash if this repo was pulled as is. Or is that directory to be filled in with our processed charts?

EDIT: I think I’m understanding what this is doing a little better. I think we should make a sub directory for all the processed charts, so the route would be more like `charts/<chartName>` or something.

ianfisk commented:
@Walker60 this looks like a pretty sweet start 👍 
I’ll give others a chance to review and then we could probably merge this...

Walker60 committed:
Add files via upload

ianfisk commented on Downloader.py:
```diff
@@ -0,0 +1,108 @@
+        temp.append(row)
+
+    if not len(database) == len(temp):
+        print "Database and Temp database do not match"
```
Should we return in this case? Or does it handle merging the two dbs. Ie basically doing a set intersect operation and storing that as the new db. If we don't handle merging we may lose data, or we need to decide which dh to prefer

ianfisk commented on Downloader.py:
```diff
@@ -0,0 +1,108 @@
+
+    if not os.path.exists(database_path):
+        print "databse file not found"
+        return
```
Should we be sending an email at this point? This seems bad haha

ianfisk commented on database:
```diff
@@ -0,0 +1,54 @@
+Albuquerque,98,Albuquerque
```
I think this database should be storing key value pairs. Right now this is fine, but we know we will be adding more information to this to populate the models in the app, and any change in the ordering of the entries in these lines will break the above code. We should try to stay away from _magic_ indexes that get us a specific piece of our model (ie the version number).

If we stored the database as key value pairs in JSON that can be mapped to dictionaries like:

```
{
   "anchorage": {
       "version": 98,
       "folderName": "anchorage"
   }
   "atlanta": {...}
   ...
}
```

Then you can easily access rows after mapping into memory like `entry = database["anchorage"]` and you can then access the version number like `version = entry["version"]`. Then we can move away from `row1[2]` stuff since you can't tell what `row1[2]` is without looking at the database, AND if we add new variables to our model, we won't break existing code and we wouldn't be restricted to only adding new variables to the end of the line.

ianfisk commented on Downloader.py:
```diff
@@ -0,0 +1,108 @@
+
+
+def downloader():
+    database_path = "/home/chris/iCharts/iChartsServices/data/database"
```
Probably use some sort of relative path to this git directory because this will break on our machines.

ianfisk commented on Downloader.py:
```diff
@@ -0,0 +1,108 @@
+
+def downloader():
+    database_path = "/home/chris/iCharts/iChartsServices/data/database"
+    temp_path = "/home/chris/iCharts/Tools/WebScraper/temp"
```
Maybe the web scraper should write key value pairs like I describe above so mapping into memory is more robust and easier to read.

ianfisk commented on Downloader.py:
```diff
@@ -0,0 +1,108 @@
+
+        database.append(row)
+
+    for line in temp_file:
```
Then if we convert to key-value pairs we could probably use a JSON parser making this cleaner

ianfisk commented on Downloader.py:
```diff
@@ -0,0 +1,108 @@
+            for files in os.listdir(directory):
+                if files.endswith(".tif"):
+                    os.rename(directory + "/" + files, directory + "/" + filename + ".tif")
+                    call(["gdal_translate", "-of","vrt","-expand", "rgba",  directory + "/" + filename +  ".tif",  directory + "/temp.vrt"])
```
Probably should check return codes to make sure everything went ok

ianfisk commented on server.py:
```diff
@@ -0,0 +1,23 @@
+@app.route("/version", methods=['GET'])
+def database():
+    try:
+        return send_from_directory(filename="database")
```
Especially since we will be expecting JSON in the app, we need to store the db as JSON as described above

ianfisk commented on Downloader.py:
```diff
@@ -0,0 +1,108 @@
+    else:
+        print "no charts to update"
+
+downloader()
```
When is this script run? And what's the flow of our system? I'm kind of gathering it is something like:
1) web scraper runs every day or something.
2) generates a temporary database
3) This downloader is script is executed XX hours/days/? and by who to update our server's database.
4) Our server disposes it's in-memory database every XX hours/days/? and reads in the updated db.
5) App gets new info.

Is that what you were thinking?

I'm worried about who owns this downloader process. Maybe the flow should be like
1) web scraper runs, generates temp db with all necessary model info.
2) web scraper runs this downloader script to compare the server's db to its own and download/tile new versions. 
3) I think what Clauson was saying is good: Have this downloader generate a new folder that contains (a) a file that holds all the information needed for the model (version number, dates, etc) and (b) the zipped folder of tiles.
4) Then the server notices that a folder has been added or changed (maybe with something like [this](https://pypi.python.org/pypi/watchdog)) and then opens up the file storing the model information, and replaces that line in its db with the new info. Or the web scraper could hit a route that tells the server to update its db, which would induce a directory scan to look for new directories or something.

Walker60 commented:
Yeah, I totally ment to commit these to my fork, not the main one. I'm
still playing with them

On Nov 4, 2016 9:17 AM, "Ian Fisk" notifications@github.com wrote:

> ## _@ianfisk_ commented on this pull request.
> 
> In Downloader.py
> https://github.com/WWU-iCharts/iChartsServices/pull/1#pullrequestreview-7207442
> :
> 
> > -        row.append(split[2].strip())
> >   +
> > -        database.append(row)
> >   +
> > -    for line in temp_file:
> > -        split = line.split(',')
> >   +
> > -        row = []
> > -        row.append(split[0])
> > -        row.append(split[1])
> > -        row.append(split[2].strip())
> >   +
> > -        temp.append(row)
> >   +
> > -    if not len(database) == len(temp):
> > -        print "Database and Temp database do not match"
> 
> Should we return in this case? Or does it handle merging the two dbs. Ie
> basically doing a set intersect operation and storing that as the new db.
> If we don't handle merging we may lose data, or we need to decide which dh
> 
> ## to prefer
> 
> In Downloader.py
> https://github.com/WWU-iCharts/iChartsServices/pull/1#pullrequestreview-7207442
> :
> 
> > +import os.path
> > +import urllib
> > +import zipfile
> > +from subprocess import call
> > +import shutil
> > +
> > +
> > +def downloader():
> > -    database_path = "/home/chris/iCharts/iChartsServices/data/database"
> > -    temp_path = "/home/chris/iCharts/Tools/WebScraper/temp"
> > -    chart_path = "/home/chris/iCharts/iChartsServices/"
> > -    did_update = False
> >   +
> > -    if not os.path.exists(database_path):
> > -        print "databse file not found"
> > -        return
> 
> ## Should we be sending an email at this point? This seems bad haha
> 
> In database
> https://github.com/WWU-iCharts/iChartsServices/pull/1#pullrequestreview-7207442
> :
> 
> > @@ -0,0 +1,54 @@
> > +Albuquerque,98,Albuquerque
> 
> I think this database should be storing key value pairs. Right now this is
> fine, but we know we will be adding more information to this to populate
> the models in the app, and any change in the ordering of the entries in
> these lines will break the above code. We should try to stay away from
> _magic_ indexes that get us a specific piece of our model (ie the version
> number).
> 
> If we stored the database as key value pairs in JSON that can be mapped to
> dictionaries like:
> 
> {
>    "anchorage": {
>        "version": 98,
>        "folderName": "anchorage"
>    }
>    "atlanta": {...}
>    ...
> }
> 
> Then you can easily access rows after mapping into memory like entry =
> database["anchorage"] and you can then access the version number like version
> = entry["version"]. Then we can move away from row1[2] stuff since you
> can't tell what row1[2] is without looking at the database, AND if we add
> new variables to our model, we won't break existing code and we wouldn't be
> 
> ## restricted to only adding new variables to the end of the line.
> 
> In Downloader.py
> https://github.com/WWU-iCharts/iChartsServices/pull/1#pullrequestreview-7207442
> :
> 
> > @@ -0,0 +1,108 @@
> > +import os.path
> > +import urllib
> > +import zipfile
> > +from subprocess import call
> > +import shutil
> > +
> > +
> > +def downloader():
> > -    database_path = "/home/chris/iCharts/iChartsServices/data/database"
> 
> Probably use some sort of relative path to this git directory because this
> 
> ## will break on our machines.
> 
> In Downloader.py
> https://github.com/WWU-iCharts/iChartsServices/pull/1#pullrequestreview-7207442
> :
> 
> > @@ -0,0 +1,108 @@
> > +import os.path
> > +import urllib
> > +import zipfile
> > +from subprocess import call
> > +import shutil
> > +
> > +
> > +def downloader():
> > -    database_path = "/home/chris/iCharts/iChartsServices/data/database"
> > -    temp_path = "/home/chris/iCharts/Tools/WebScraper/temp"
> 
> Maybe the web scraper should write key value pairs like I describe above
> 
> ## so mapping into memory is more robust and easier to read.
> 
> In Downloader.py
> https://github.com/WWU-iCharts/iChartsServices/pull/1#pullrequestreview-7207442
> :
> 
> > +
> > -    database = []
> > -    temp = []
> >   +
> > -    #Read data and put into two arrays
> > -    for line in database_file:
> > -        split = line.split(',')
> >   +
> > -        row = []
> > -        row.append(split[0])
> > -        row.append(split[1])
> > -        row.append(split[2].strip())
> >   +
> > -        database.append(row)
> >   +
> > -    for line in temp_file:
> 
> Then if we convert to key-value pairs we could probably use a JSON parser
> 
> ## making this cleaner
> 
> In Downloader.py
> https://github.com/WWU-iCharts/iChartsServices/pull/1#pullrequestreview-7207442
> :
> 
> > -            urllib.urlretrieve (row2[2], row1[2] +".zip")
> >   +
> > -            #unzip file
> > -            print "unzipping"
> > -            zip_ref = zipfile.ZipFile(row1[2]+ ".zip", 'r')
> > -            zip_ref.extractall(row1[2])
> > -            zip_ref.close()
> >   +
> > -            #Tile map
> > -            print "Creating tiles"
> > -            directory = row1[2]
> > -            filename = row1[2]
> > -            for files in os.listdir(directory):
> > -                if files.endswith(".tif"):
> > -                    os.rename(directory + "/" + files, directory + "/" + filename + ".tif")
> > -                    call(["gdal_translate", "-of","vrt","-expand", "rgba",  directory + "/" + filename +  ".tif",  directory + "/temp.vrt"])
> 
> ## Probably should check return codes to make sure everything went ok
> 
> In server.py
> https://github.com/WWU-iCharts/iChartsServices/pull/1#pullrequestreview-7207442
> :
> 
> > @@ -0,0 +1,23 @@
> > +from flask import Flask, jsonify
> > +from flask import send_from_directory
> > +
> > +app = Flask(**name**)
> > +
> > +
> > +@app.route("/version", methods=['GET'])
> > +def database():
> > -    try:
> > -        return send_from_directory(filename="database")
> 
> Especially since we will be expecting JSON in the app, we need to store
> 
> ## the db as JSON as described above
> 
> In Downloader.py
> https://github.com/WWU-iCharts/iChartsServices/pull/1#pullrequestreview-7207442
> :
> 
> > -            os.rename(row1[2] + ".zip", chart_path + row1[2] + "/" + row1[2] + ".zip")
> >   +
> > -            #clean up
> > -            print "cleaning up"
> > -            shutil.rmtree(directory)
> > -            did_update = True
> >   +
> > -        updated_file.write(row1[0] + ',' + row1[1] + ',' + row1[2] + '\n')
> >   +
> > -    if did_update:
> > -        print "updating database"
> > -        #os.rename("updated", database_path)
> > -    else:
> > -        print "no charts to update"
> >   +
> >   +downloader()
> 
> When is this script run? And what's the flow of our system? I'm kind of
> gathering it is something like:
> 1) web scraper runs every day or something.
> 2) generates a temporary database
> 3) This downloader is script is executed XX hours/days/? and by who to
> update our server's database.
> 4) Our server disposes it's in-memory database every XX hours/days/? and
> reads in the updated db.
> 5) App gets new info.
> 
> Is that what you were thinking?
> 
> I'm worried about who owns this downloader process. Maybe the flow should
> be like
> 1) web scraper runs, generates temp db with all necessary model info.
> 2) web scraper runs this downloader script to compare the server's db to
> its own and download/tile new versions.
> 3) I think what Clauson was saying is good: Have this downloader generate
> a new folder that contains (a) a file that holds all the information needed
> for the model (version number, dates, etc) and (b) the zipped folder of
> tiles.
> 4) Then the server notices that a folder has been added or changed (maybe
> with something like this https://pypi.python.org/pypi/watchdog) and
> then opens up the file storing the model information, and replaces that
> line in its db with the new info. Or the web scraper could hit a route that
> tells the server to update its db, which would induce a directory scan to
> look for new directories or something.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/WWU-iCharts/iChartsServices/pull/1#pullrequestreview-7207442,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AQPsJ74ijfEk14rDCglqm_BWGpMPEUH5ks5q61qigaJpZM4KfPfU
> .

ianfisk commented:
@Walker60 no these are on your fork. Basically after you PR, anything else you push to that branch is automatically shown here for better or worse haha
